In [1]:
# Imports
import pandas as pd
import json
from pandas import json_normalize
import numpy as np
# Configurações para melhor visualização dos dados
pd.set_option('display.max_columns',100)
pd.set_option('display.max_colwidth', None)

Entender quais informações o conjunto de dados possui

In [2]:
# Carregar dados repositório git do desafio

df = pd.read_json('https://raw.githubusercontent.com/SilvioCSOliveira/ChallengeDataScience2022AluraVoz/main/Telco-Customer-Churn.json')
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7267 entries, 0 to 7266
Data columns (total 6 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   customerID  7267 non-null   object
 1   Churn       7267 non-null   object
 2   customer    7267 non-null   object
 3   phone       7267 non-null   object
 4   internet    7267 non-null   object
 5   account     7267 non-null   object
dtypes: object(6)
memory usage: 340.8+ KB


In [3]:
# Visualizar conteúdo do arquivo json 
df.head()

customerID Churn  \
0  0002-ORFBO    No   
1  0003-MKNFE    No   
2  0004-TLHLJ   Yes   
3  0011-IGKFF   Yes   
4  0013-EXCHZ   Yes   

                                                                                       customer  \
0  {'gender': 'Female', 'SeniorCitizen': 0, 'Partner': 'Yes', 'Dependents': 'Yes', 'tenure': 9}   
1      {'gender': 'Male', 'SeniorCitizen': 0, 'Partner': 'No', 'Dependents': 'No', 'tenure': 9}   
2      {'gender': 'Male', 'SeniorCitizen': 0, 'Partner': 'No', 'Dependents': 'No', 'tenure': 4}   
3    {'gender': 'Male', 'SeniorCitizen': 1, 'Partner': 'Yes', 'Dependents': 'No', 'tenure': 13}   
4   {'gender': 'Female', 'SeniorCitizen': 1, 'Partner': 'Yes', 'Dependents': 'No', 'tenure': 3}   

                                             phone  \
0   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
1  {'PhoneService': 'Yes', 'MultipleLines': 'Yes'}   
2   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
3   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
4   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   

                                                                                                                                                                            internet  \
0           {'InternetService': 'DSL', 'OnlineSecurity': 'No', 'OnlineBackup': 'Yes', 'DeviceProtection': 'No', 'TechSupport': 'Yes', 'StreamingTV': 'Yes', 'StreamingMovies': 'No'}   
1             {'InternetService': 'DSL', 'OnlineSecurity': 'No', 'OnlineBackup': 'No', 'DeviceProtection': 'No', 'TechSupport': 'No', 'StreamingTV': 'No', 'StreamingMovies': 'Yes'}   
2     {'InternetService': 'Fiber optic', 'OnlineSecurity': 'No', 'OnlineBackup': 'No', 'DeviceProtection': 'Yes', 'TechSupport': 'No', 'StreamingTV': 'No', 'StreamingMovies': 'No'}   
3  {'InternetService': 'Fiber optic', 'OnlineSecurity': 'No', 'OnlineBackup': 'Yes', 'DeviceProtection': 'Yes', 'TechSupport': 'No', 'StreamingTV': 'Yes', 'StreamingMovies': 'Yes'}   
4    {'InternetService': 'Fiber optic', 'OnlineSecurity': 'No', 'OnlineBackup': 'No', 'DeviceProtection': 'No', 'TechSupport': 'Yes', 'StreamingTV': 'Yes', 'StreamingMovies': 'No'}   

                                                                                                                                            account  
0              {'Contract': 'One year', 'PaperlessBilling': 'Yes', 'PaymentMethod': 'Mailed check', 'Charges': {'Monthly': 65.6, 'Total': '593.3'}}  
1         {'Contract': 'Month-to-month', 'PaperlessBilling': 'No', 'PaymentMethod': 'Mailed check', 'Charges': {'Monthly': 59.9, 'Total': '542.4'}}  
2   {'Contract': 'Month-to-month', 'PaperlessBilling': 'Yes', 'PaymentMethod': 'Electronic check', 'Charges': {'Monthly': 73.9, 'Total': '280.85'}}  
3  {'Contract': 'Month-to-month', 'PaperlessBilling': 'Yes', 'PaymentMethod': 'Electronic check', 'Charges': {'Monthly': 98.0, 'Total': '1237.85'}}  
4        {'Contract': 'Month-to-month', 'PaperlessBilling': 'Yes', 'PaymentMethod': 'Mailed check', 'Charges': {'Monthly': 83.9, 'Total': '267.4'}}

In [4]:
# 'serializar' campos que possuem jsons no seu conteúdo,  customer phone internet e account
df_customer = pd.json_normalize(data = df.customer) 
df_phone = pd.json_normalize(data = df.phone) 
df_internet = pd.json_normalize(data = df.internet) 
df_account = pd.json_normalize(data = df.account) 

In [5]:
# 'Juntar' campos jsons 'serializados' e retirar campos que possuem jsons no seu conteúdo
df_telco_customer = pd.concat([df, df_customer,df_phone,df_internet,df_account], axis = 1)
df_telco_customer = df_telco_customer.drop(columns = ['phone','customer','internet','account'])

df_telco_customer.head(1)

customerID Churn  gender  SeniorCitizen Partner Dependents  tenure  \
0  0002-ORFBO    No  Female              0     Yes        Yes       9   

  PhoneService MultipleLines InternetService OnlineSecurity OnlineBackup  \
0          Yes            No             DSL             No          Yes   

  DeviceProtection TechSupport StreamingTV StreamingMovies  Contract  \
0               No         Yes         Yes              No  One year   

  PaperlessBilling PaymentMethod  Charges.Monthly Charges.Total  
0              Yes  Mailed check             65.6         593.3

Traduzir as colunas

In [6]:
df_telco_customer.rename(columns={'customerID' : 'IdCliente',	
'Churn': 'ClienteCancelado',
'gender' : 'Genero'	,
'SeniorCitizen' : 'Maior65',
'Partner' : 'Parceiro(a)',
'Dependents': 'Dependentes',
'tenure' : 'TempoContrato(m)',
'PhoneService' : 'ServiçoTelefônico',
'MultipleLines' : 'MultiplasLinhas',
'InternetService' : 'ServicoInternet',
'OnlineSecurity' : 'AdicinalSegurancaOnline',
'OnlineBackup' : 'AdicionalBackup',
'DeviceProtection' : 'AdicionalProtDispositivo',
'TechSupport' : 'AdicionalSuporte',
'StreamingTV': 'AssinaturaTv',
'StreamingMovies' : 'AssinaturaFilmes',
'Contract' : 'TipoContrato',
'PaperlessBilling': 'FaturaDigital',
'PaymentMethod': 'FormaPagamento',
'Charges.Monthly': 'GastoMensal' ,
'Charges.Total': 'GastoTotal'}, inplace=True)

df_telco_customer.head(5)


IdCliente ClienteCancelado  Genero  Maior65 Parceiro(a) Dependentes  \
0  0002-ORFBO               No  Female        0         Yes         Yes   
1  0003-MKNFE               No    Male        0          No          No   
2  0004-TLHLJ              Yes    Male        0          No          No   
3  0011-IGKFF              Yes    Male        1         Yes          No   
4  0013-EXCHZ              Yes  Female        1         Yes          No   

   TempoContrato(m) ServiçoTelefônico MultiplasLinhas ServicoInternet  \
0                 9               Yes              No             DSL   
1                 9               Yes             Yes             DSL   
2                 4               Yes              No     Fiber optic   
3                13               Yes              No     Fiber optic   
4                 3               Yes              No     Fiber optic   

  AdicinalSegurancaOnline AdicionalBackup AdicionalProtDispositivo  \
0                      No             Yes                       No   
1                      No              No                       No   
2                      No              No                      Yes   
3                      No             Yes                      Yes   
4                      No              No                       No   

  AdicionalSuporte AssinaturaTv AssinaturaFilmes    TipoContrato  \
0              Yes          Yes               No        One year   
1               No           No              Yes  Month-to-month   
2               No           No               No  Month-to-month   
3               No          Yes              Yes  Month-to-month   
4              Yes          Yes               No  Month-to-month   

  FaturaDigital    FormaPagamento  GastoMensal GastoTotal  
0           Yes      Mailed check         65.6      593.3  
1            No      Mailed check         59.9      542.4  
2           Yes  Electronic check         73.9     280.85  
3           Yes  Electronic check         98.0    1237.85  
4           Yes      Mailed check         83.9      267.4

Conhecer o conteúdo das colunas, verificar quais são as inconsistências nos dados e  Corrigir as inconsistências nos dados.


 Dicionário de dados:
- IdCliente                 : Número de identificação único de cada cliente
- ClienteCancelado          : Se o cliente deixou ou não a empresa
- Genero                    : Gênero (masculino e feminino)
- Maior65                   : Informação sobre um cliente ter ou não idade igual ou maior que 65 anos
- Parceiro(a)               : Se o cliente possui ou não um parceiro ou parceira
- Dependentes               : Se o cliente possui ou não dependentes
- TempoContrato(m)          : Meses de contrato do cliente
- ServiçoTelefônico         : Assinatura de serviço telefônico
- MultiplasLinhas           : Assisnatura de mais de uma linha de telefone
- ServicoInternet           : Assinatura de um provedor internet
- AdicinalSegurancaOnline   : Assinatura adicional de segurança online
- AdicionalBackup           : Assinatura adicional de backup online
- AdicionalProtDispositivo  : Assinatura adicional de proteção no dispositivo
- AdicionalSuporte          : Assinatura adicional de suporte técnico, menos tempo de espera
- AssinaturaTv              : Assinatura de TV a cabo
- AssinaturaFilmes          : Assinatura de streaming de filmes
- TipoContrato              : Tipo de contrato
- FaturaDigital             : Se o cliente prefere receber online a fatura
- FormaPagamento            : Forma de pagamento
- GastoMensal               : Total de todos os serviços do cliente por mês
- GastoTotal                : Total gasto pelo cliente


In [7]:

df_telco_customer.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7267 entries, 0 to 7266
Data columns (total 21 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   IdCliente                 7267 non-null   object 
 1   ClienteCancelado          7267 non-null   object 
 2   Genero                    7267 non-null   object 
 3   Maior65                   7267 non-null   int64  
 4   Parceiro(a)               7267 non-null   object 
 5   Dependentes               7267 non-null   object 
 6   TempoContrato(m)          7267 non-null   int64  
 7   ServiçoTelefônico         7267 non-null   object 
 8   MultiplasLinhas           7267 non-null   object 
 9   ServicoInternet           7267 non-null   object 
 10  AdicinalSegurancaOnline   7267 non-null   object 
 11  AdicionalBackup           7267 non-null   object 
 12  AdicionalProtDispositivo  7267 non-null   object 
 13  AdicionalSuporte          7267 non-null   object 
 14  Assinatu

In [8]:
# Verificando o conteúdo dos campos
df_telco_customer['IdCliente'].value_counts()

0002-ORFBO    1
6614-VBEGU    1
6637-KYRCV    1
6635-MYYYZ    1
6635-CPNUN    1
             ..
3374-TTZTK    1
3374-PZLXD    1
3374-LXDEV    1
3373-YZZYM    1
9995-HOTOH    1
Name: IdCliente, Length: 7267, dtype: int64

In [9]:
df_telco_customer['ClienteCancelado'].value_counts()
# 224 nulos, necessário corrigir antes de iniciar a análise exploratória, o objetivo do estudo é analisar o motivo dos cancelamentos e dos não cancelamentos

No     5174
Yes    1869
        224
Name: ClienteCancelado, dtype: int64

In [10]:
df_telco_customer['Genero'].value_counts()
# Ok

Male      3675
Female    3592
Name: Genero, dtype: int64

In [11]:
df_telco_customer['Maior65'].value_counts()
# Aparentemente a coluna indica que o assinante titular é Maior65 = 1 e não é = 0

0    6085
1    1182
Name: Maior65, dtype: int64

In [12]:
df_telco_customer['Parceiro(a)'].value_counts()
# Ok

No     3749
Yes    3518
Name: Parceiro(a), dtype: int64

In [13]:
df_telco_customer['Dependentes'].value_counts()
# Ok

No     5086
Yes    2181
Name: Dependentes, dtype: int64

In [14]:
df_telco_customer['TempoContrato(m)'].value_counts().sort_index()
# Clientes com tempo de contrato igual a zero '0', significa que e um cliente novo que ainda não completou o primeiro mês de contrato

0      11
1     634
2     246
3     207
4     185
     ... 
68    104
69    100
70    122
71    176
72    369
Name: TempoContrato(m), Length: 73, dtype: int64

In [15]:
df_telco_customer['ServiçoTelefônico'].value_counts()
# Ok

Yes    6560
No      707
Name: ServiçoTelefônico, dtype: int64

Informação importante 707 **não possuem serviço de telefonico**, será que isso influencia o cancelamento ?

In [16]:
df_telco_customer['MultiplasLinhas'].value_counts()
# Ok

No                  3495
Yes                 3065
No phone service     707
Name: MultiplasLinhas, dtype: int64

In [17]:
df_telco_customer['ServicoInternet'].value_counts()
# Ok
# Informação importante 1581 não possuem serviço de internet

Fiber optic    3198
DSL            2488
No             1581
Name: ServicoInternet, dtype: int64

Informação importante 1581 **não possuem serviço de internet**, será que isso influencia o cancelamento ?

In [18]:
df_telco_customer['AdicinalSegurancaOnline'].value_counts()
# Ok

No                     3608
Yes                    2078
No internet service    1581
Name: AdicinalSegurancaOnline, dtype: int64

In [19]:
df_telco_customer['AdicionalBackup'].value_counts()
# Ok

No                     3182
Yes                    2504
No internet service    1581
Name: AdicionalBackup, dtype: int64

In [20]:
df_telco_customer['AdicionalProtDispositivo'].value_counts()
# Ok

No                     3195
Yes                    2491
No internet service    1581
Name: AdicionalProtDispositivo, dtype: int64

In [21]:
df_telco_customer['AdicionalSuporte'].value_counts()
# Ok

No                     3582
Yes                    2104
No internet service    1581
Name: AdicionalSuporte, dtype: int64

In [22]:
df_telco_customer['AssinaturaTv'].value_counts()
# Ok

No                     2896
Yes                    2790
No internet service    1581
Name: AssinaturaTv, dtype: int64

Informação importante 2896 não possuem assinatura de TV, será que isso influencia o cancelamento ?

In [23]:
df_telco_customer['AssinaturaFilmes'].value_counts()
# Ok

No                     2870
Yes                    2816
No internet service    1581
Name: AssinaturaFilmes, dtype: int64

In [24]:
df_telco_customer['TipoContrato'].value_counts()
# Ok

Month-to-month    4005
Two year          1743
One year          1519
Name: TipoContrato, dtype: int64

In [25]:
df_telco_customer['FaturaDigital'].value_counts()
# Ok

Yes    4311
No     2956
Name: FaturaDigital, dtype: int64

In [26]:
df_telco_customer['FormaPagamento'].value_counts()
# Ok

Electronic check             2445
Mailed check                 1665
Bank transfer (automatic)    1589
Credit card (automatic)      1568
Name: FormaPagamento, dtype: int64

In [27]:
df_telco_customer['GastoMensal'].value_counts().sort_index()

18.25     1
18.40     1
18.55     1
18.70     2
18.75     2
         ..
118.20    1
118.35    1
118.60    2
118.65    1
118.75    1
Name: GastoMensal, Length: 1585, dtype: int64

In [28]:
df_telco_customer['GastoTotal'].value_counts().sort_index()
# Ok
# Os 11 clientes que não possuem valor de gasto total são os 11 clientes que ainda não tem completaram o primeiro mês de contrato (campo 'TempoContrato(m)')

          11
100.2      1
100.25     1
100.35     1
100.4      1
          ..
997.75     1
998.1      1
999.45     2
999.8      1
999.9      1
Name: GastoTotal, Length: 6531, dtype: int64

In [29]:
# Como o campo GastoTotal representa valor monetário, vou substituir o valor dos clientes que ainda não completaram o primeiro mês de 'vazio' para zero '0' 
df_telco_customer.loc[df_telco_customer['GastoTotal'] == ' ' , 'GastoTotal'] = '0'
df_telco_customer['GastoTotal'].value_counts().sort_index()

0         11
100.2      1
100.25     1
100.35     1
100.4      1
          ..
997.75     1
998.1      1
999.45     2
999.8      1
999.9      1
Name: GastoTotal, Length: 6531, dtype: int64

In [30]:
# Como o campo GastoTotal representa valor monetário deveria ser do tipo float64 é não object, necessário corrigir para realização das futuras análises
df_telco_customer['GastoTotal'] = pd.to_numeric(df_telco_customer['GastoTotal'],errors = 'coerce')
df_telco_customer.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7267 entries, 0 to 7266
Data columns (total 21 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   IdCliente                 7267 non-null   object 
 1   ClienteCancelado          7267 non-null   object 
 2   Genero                    7267 non-null   object 
 3   Maior65                   7267 non-null   int64  
 4   Parceiro(a)               7267 non-null   object 
 5   Dependentes               7267 non-null   object 
 6   TempoContrato(m)          7267 non-null   int64  
 7   ServiçoTelefônico         7267 non-null   object 
 8   MultiplasLinhas           7267 non-null   object 
 9   ServicoInternet           7267 non-null   object 
 10  AdicinalSegurancaOnline   7267 non-null   object 
 11  AdicionalBackup           7267 non-null   object 
 12  AdicionalProtDispositivo  7267 non-null   object 
 13  AdicionalSuporte          7267 non-null   object 
 14  Assinatu

In [31]:
# Verificando a transformação do campo
df_telco_customer['GastoTotal'].value_counts().sort_index()

0.00       11
18.80       1
18.85       2
18.90       2
19.00       1
           ..
8564.75     1
8594.40     1
8670.10     1
8672.45     1
8684.80     1
Name: GastoTotal, Length: 6531, dtype: int64

In [32]:
# Como o objetivo é analisar o cancelamento (churn), vou excluir os 224 clientes sem informação de cancelamento
index = df_telco_customer[ df_telco_customer['ClienteCancelado'] == ''].index
df_telco_customer.drop(index , inplace=True)
df_telco_customer['ClienteCancelado'].value_counts()

No     5174
Yes    1869
Name: ClienteCancelado, dtype: int64

In [33]:
# Como somente a coluna 'Maior65' (SeniorCitizen) não está seguindo o padrão de conteudo para domínios boleanos da massa de dados, ao invés de  'Yes' e 'No' está usando 1 e 0, 
# decidi, ajusta-la para o mesmo padrão.
map =  {1:'Yes', 0:'No'}
df_telco_customer['Maior65'].replace(map, inplace=True)
df_telco_customer.head(5)


IdCliente ClienteCancelado  Genero Maior65 Parceiro(a) Dependentes  \
0  0002-ORFBO               No  Female      No         Yes         Yes   
1  0003-MKNFE               No    Male      No          No          No   
2  0004-TLHLJ              Yes    Male      No          No          No   
3  0011-IGKFF              Yes    Male     Yes         Yes          No   
4  0013-EXCHZ              Yes  Female     Yes         Yes          No   

   TempoContrato(m) ServiçoTelefônico MultiplasLinhas ServicoInternet  \
0                 9               Yes              No             DSL   
1                 9               Yes             Yes             DSL   
2                 4               Yes              No     Fiber optic   
3                13               Yes              No     Fiber optic   
4                 3               Yes              No     Fiber optic   

  AdicinalSegurancaOnline AdicionalBackup AdicionalProtDispositivo  \
0                      No             Yes                       No   
1                      No              No                       No   
2                      No              No                      Yes   
3                      No             Yes                      Yes   
4                      No              No                       No   

  AdicionalSuporte AssinaturaTv AssinaturaFilmes    TipoContrato  \
0              Yes          Yes               No        One year   
1               No           No              Yes  Month-to-month   
2               No           No               No  Month-to-month   
3               No          Yes              Yes  Month-to-month   
4              Yes          Yes               No  Month-to-month   

  FaturaDigital    FormaPagamento  GastoMensal  GastoTotal  
0           Yes      Mailed check         65.6      593.30  
1            No      Mailed check         59.9      542.40  
2           Yes  Electronic check         73.9      280.85  
3           Yes  Electronic check         98.0     1237.85  
4           Yes      Mailed check         83.9      267.40

Criar coluna de valor média de gasto diário 

In [34]:
# Para criar a coluna de gasto diário médio usei a coluna GastoMensal dividido por 30 dias
df_telco_customer = df_telco_customer.assign(GastoDiarioMedio = df_telco_customer['GastoMensal']/30)
df_telco_customer.head(5)

IdCliente ClienteCancelado  Genero Maior65 Parceiro(a) Dependentes  \
0  0002-ORFBO               No  Female      No         Yes         Yes   
1  0003-MKNFE               No    Male      No          No          No   
2  0004-TLHLJ              Yes    Male      No          No          No   
3  0011-IGKFF              Yes    Male     Yes         Yes          No   
4  0013-EXCHZ              Yes  Female     Yes         Yes          No   

   TempoContrato(m) ServiçoTelefônico MultiplasLinhas ServicoInternet  \
0                 9               Yes              No             DSL   
1                 9               Yes             Yes             DSL   
2                 4               Yes              No     Fiber optic   
3                13               Yes              No     Fiber optic   
4                 3               Yes              No     Fiber optic   

  AdicinalSegurancaOnline AdicionalBackup AdicionalProtDispositivo  \
0                      No             Yes                       No   
1                      No              No                       No   
2                      No              No                      Yes   
3                      No             Yes                      Yes   
4                      No              No                       No   

  AdicionalSuporte AssinaturaTv AssinaturaFilmes    TipoContrato  \
0              Yes          Yes               No        One year   
1               No           No              Yes  Month-to-month   
2               No           No               No  Month-to-month   
3               No          Yes              Yes  Month-to-month   
4              Yes          Yes               No  Month-to-month   

  FaturaDigital    FormaPagamento  GastoMensal  GastoTotal  GastoDiarioMedio  
0           Yes      Mailed check         65.6      593.30          2.186667  
1            No      Mailed check         59.9      542.40          1.996667  
2           Yes  Electronic check         73.9      280.85          2.463333  
3           Yes  Electronic check         98.0     1237.85          3.266667  
4           Yes      Mailed check         83.9      267.40          2.796667

In [35]:
df_telco_customer.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 7043 entries, 0 to 7266
Data columns (total 22 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   IdCliente                 7043 non-null   object 
 1   ClienteCancelado          7043 non-null   object 
 2   Genero                    7043 non-null   object 
 3   Maior65                   7043 non-null   object 
 4   Parceiro(a)               7043 non-null   object 
 5   Dependentes               7043 non-null   object 
 6   TempoContrato(m)          7043 non-null   int64  
 7   ServiçoTelefônico         7043 non-null   object 
 8   MultiplasLinhas           7043 non-null   object 
 9   ServicoInternet           7043 non-null   object 
 10  AdicinalSegurancaOnline   7043 non-null   object 
 11  AdicionalBackup           7043 non-null   object 
 12  AdicionalProtDispositivo  7043 non-null   object 
 13  AdicionalSuporte          7043 non-null   object 
 14  Assinatu

In [36]:
# Salvar os dados tratados em um CSV para usar na semana 2 
df_telco_customer.to_csv('df_telco_customer.csv',sep=';')